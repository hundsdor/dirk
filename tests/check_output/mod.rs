use std::{path::PathBuf, process::Command};

use insta::{assert_snapshot, with_settings, Settings};
use tempdir::TempDir;

pub(crate) fn test_main(path: &str, name: &str) {
    let mut cmd = Command::new(env!("CARGO"));

    cmd.arg("run");
    cmd.arg("--example");
    cmd.arg(name);

    cmd.env("RUSTFLAGS", "--cap-lints=allow");
    cmd.env("CARGO_TERM_COLOR", "never");

    let mut current_dir = PathBuf::new();
    current_dir.push("test-data");
    current_dir.push(path);

    cmd.current_dir(&current_dir);

    let target_dir = TempDir::new_in(env!("CARGO_TARGET_TMPDIR"), name).unwrap();
    cmd.env("TARGET_DIR", target_dir.path());

    let output = cmd
        .output()
        .unwrap_or_else(|_| panic!("Failed to execute test {name}"));

    let mut settings = Settings::clone_current();

    let filter_block_package_cache = r".*Blocking waiting for file lock on package cache.*\n";
    settings.add_filter(filter_block_package_cache, "");

    let filter_block_build_directory = r".*Blocking waiting for file lock on build directory.*\n";
    settings.add_filter(filter_block_build_directory, "");

    let filter_finished_targets = r".*Finished .* target\(s\) in .*s";
    settings.add_filter(filter_finished_targets, "Finished compiling target(s)");

    let filter_compiling = r".*Compiling .*\n";
    settings.add_filter(filter_compiling, "");

    let filter_updating_index = r".*Updating crates.io index\n";
    settings.add_filter(filter_updating_index, "");

    let filter_locking = r".*Locking \d+ packages to latest compatible versions\n";
    settings.add_filter(filter_locking, "");

    let filter_adding = r".*Adding .* (\(.*\))?\n";
    settings.add_filter(filter_adding, "");

    let filter_path = if cfg!(windows) {
        r"(?:([\w\-_]+)(?:\\))+([\w\-_]+?\.[\w\-_]+)?"
    } else {
        r"(?:([\w\-_]+)(?:/))+([\w\-_]+?\.[\w\-_]+)?"
    };
    settings.add_filter(filter_path, "$1/$2");

    if cfg!(windows) {
        // `\\` has already been replaced by `/`
        let filter_exe = r"Running `(?:([\w\-_]+)(?:/))+([\w\-_]+?).exe`";
        settings.add_filter(filter_exe, "Running `$1/$2`");
    }

    let pretty = format!(
        "Stdout:\n{}\n\nStderr:\n{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    );

    settings.bind(|| {
        with_settings!({snapshot_suffix => name}, {
            assert_snapshot!(pretty);
        });
    });
}
