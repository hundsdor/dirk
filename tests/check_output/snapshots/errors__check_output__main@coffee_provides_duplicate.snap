---
source: tests/check_output/mod.rs
expression: pretty
---
Stdout:


Stderr:
error[E0428]: the name `ScopedFactoryElectricHeater` is defined multiple times
   --> examples/provides_duplicate.rs:131:5
    |
121 |     #[provides(scoped_inject)]
    |     -------------------------- previous definition of the type `ScopedFactoryElectricHeater` here
...
131 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `ScopedFactoryElectricHeater` redefined here
    |
    = note: `ScopedFactoryElectricHeater` must be defined only once in the type namespace of this module
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `Provider<Rc<RefCell<ElectricHeater>>>` for type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:131:5
    |
121 |     #[provides(scoped_inject)]
    |     -------------------------- first implementation here
...
131 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `heater::ScopedFactoryElectricHeater`
    |
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0592]: duplicate definitions with name `new`
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definitions for `new`
...
131 |     #[provides(scoped_inject)]
    |     -------------------------- other definition for `new`
    |
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0592]: duplicate definitions with name `create`
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definitions for `create`
...
131 |     #[provides(scoped_inject)]
    |     -------------------------- other definition for `create`
    |
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0592]: duplicate definitions with name `new_instance`
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definitions for `new_instance`
...
131 |     #[provides(scoped_inject)]
    |     -------------------------- other definition for `new_instance`
    |
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0034]: multiple applicable items in scope
   --> examples/provides_duplicate.rs:35:25
    |
35  |     heater: scoped_bind(ElectricHeater) [logger],
    |                         ^^^^^^^^^^^^^^ multiple `create` found
    |
note: candidate #1 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:131:5
    |
131 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0034]: multiple applicable items in scope
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ multiple `new_instance` found
    |
note: candidate #1 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:131:5
    |
131 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0034]: multiple applicable items in scope
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ multiple `new` found
    |
note: candidate #1 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:131:5
    |
131 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0034]: multiple applicable items in scope
   --> examples/provides_duplicate.rs:131:5
    |
131 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ multiple `new_instance` found
    |
note: candidate #1 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:131:5
    |
131 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0034]: multiple applicable items in scope
   --> examples/provides_duplicate.rs:131:5
    |
131 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ multiple `new` found
    |
note: candidate #1 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:121:5
    |
121 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `heater::ScopedFactoryElectricHeater`
   --> examples/provides_duplicate.rs:131:5
    |
131 |     #[provides(scoped_inject)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the attribute macro `provides` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0034, E0119, E0428, E0592.
For more information about an error, try `rustc --explain E0034`.
error: could not compile `coffee` (example "provides_duplicate") due to 10 previous errors
